### PR TITLE
[HOPSWORKS-1303] Kafka REST API v2 - Topics

### PR DESCRIPTION
**Pull request by kai-chi**
## Make sure there is no duplicate PR for this issue

* **Please check if the PR meets the following requirements**
- [x] Adds tests for the submitted changes (for bug fixes & features)
- [x] Passes the tests
- [x] HOPSWORKS JIRA issue has been opened for this PR
- [x] All commits have been squashed down to a single commit

* **Post a link to the associated JIRA issue**
https://logicalclocks.atlassian.net/browse/HOPSWORKS-1303

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Refactoring Kafka topics API

* **What is the new behavior (if this is a feature change)?**

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
No

* **Other information**:

### CODE REVIEW

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java [hunk 1]
```diff
@@ -109,62 +103,59 @@
   @EJB
   Settings settings;
   @EJB
-  private ProjectFacade projectsFacade;
-  @EJB
-  private CertificateMaterializer certificateMaterializer;
-  @EJB
   private BaseHadoopClientsService baseHadoopService;
+  //TODO: added temporarly to compile the unrefactored endpoints - remove
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: @kai-chi reminder

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java [hunk 2]
```diff
@@ -109,62 +103,59 @@
...
-        topics.add(new TopicDTO(pt.getTopicName(),
-            pt.getSchemaTopics().getSchemaTopicsPK().getName(),
-            pt.getSchemaTopics().getSchemaTopicsPK().getVersion()));
-      }
-    }
-    return topics;
+  
+  @PostConstruct
+  private void init() {
+    adminClient = getAdminClient();
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: @kai-chi It might be better to move this adminClient to its own singleton. The pool will contain many KafkaFacade instances, do we need one adminClient for each instance?
@ErmiasG any opinion?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java [hunk 3]
```diff
@@ -173,35 +164,24 @@ public ProjectTopics findTopicByNameAndProject(Project project, String topicName
    *
    */
   public List<SharedTopics> findSharedTopicsByProject(Integer projectId) {
-    TypedQuery<SharedTopics> query = em.createNamedQuery(
-        "SharedTopics.findByProjectId",
-        SharedTopics.class);
-    query.setParameter("projectId", projectId);
-    return query.getResultList();
+    return em.createNamedQuery("SharedTopics.findByProjectId", SharedTopics.class)
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: @kai-chi instead of a list, we might need to return a CollectionInfo, as for pagination we need to return the count as well https://github.com/logicalclocks/hopsworks/blob/master/hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/jobs/description/JobFacade.java#L208 
That means KafkaFacade would extend the AbstractFacade. 

How does the solution presented here handle pagination?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java [hunk 4]
```diff
@@ -75,7 +94,297 @@
...
+  public void removeTopicFromProject(Project project, String topicName) throws KafkaException {
+    
+    ProjectTopics pt = kafkaFacade.findTopicByNameAndProject(project, topicName)
+      .orElseThrow(() ->
+        new KafkaException(RESTCodes.KafkaErrorCode.TOPIC_NOT_FOUND, Level.FINE, "topic: " + topicName));
+    
+    kafkaFacade.removeTopicFromProject(pt);
+  }
+  
+  public List<TopicDTO> findTopicDtosByProject(Project project) {
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java]: @kai-chi let's ommit the Dtos from the name

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java [hunk 5]
```diff
@@ -75,7 +94,297 @@
...
+    List<ProjectTopics> ptList = kafkaFacade.findTopicsByProject(project);
+  
+    List<TopicDTO> topics = new ArrayList<>();
+    if(ptList != null && !ptList.isEmpty()) {
+      topics = new ArrayList<>();
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java]: @kai-chi `topics` is initialized two rows above

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java [hunk 6]
```diff
@@ -75,7 +94,297 @@
...
+    }
+    
+    SchemaTopics schema =
+      kafkaFacade.findSchemaByNameAndVersion(topicDto.getSchemaName(), topicDto.getSchemaVersion()).orElseThrow(() ->
+        new KafkaException(RESTCodes.KafkaErrorCode.SCHEMA_NOT_FOUND, Level.FINE, "topic: " + topicName));
+    
+    //check if the replication factor is not greater than the
+    //number of running brokers
+    try {
+      Set<String> brokerEndpoints = settings.getBrokerEndpoints();
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java]: @kai-chi can we use `getTopicDetailsFromKafkaCluster` instead?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/util/KafkaServingHelper.java [hunk 7]
```diff
@@ -160,28 +165,29 @@ private ProjectTopics setupKafkaTopic(Project project, ServingWrapper servingWra
       Settings.INFERENCE_SCHEMAVERSION);
 
     ProjectTopics pt = null;
-    pt = kafkaFacade.createTopicInProject(project, topicDTO);
+    SchemaTopics schema =
+      kafkaFacade.findSchemaByNameAndVersion(topicDTO.getSchemaName(), topicDTO.getSchemaVersion()).orElseThrow(() ->
+        new KafkaException(RESTCodes.KafkaErrorCode.SCHEMA_NOT_FOUND, Level.FINE, "topic: " + topicDTO.getName()));
+  
+    pt = kafkaFacade.createTopicInProject(project, topicDTO, schema);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/serving/util/KafkaServingHelper.java]: @kai-chi we can make life easier for the client by having `kafkaFacade.createTopicInProject(project, topicDTO);` which does 
```
    SchemaTopics schema =
      kafkaFacade.findSchemaByNameAndVersion(topicDTO.getSchemaName(), topicDTO.getSchemaVersion()).orElseThrow(() ->
        new KafkaException(RESTCodes.KafkaErrorCode.SCHEMA_NOT_FOUND, Level.FINE, "topic: " + topicDTO.getName()));
```

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java (hunk 6 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java]: no, `getTopicDetailsFromKafkaCluster` returns details of a topic, while here you are creating a new topic

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 2 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: introduced a singleton :+1:

Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: with singletons we need to be careful in case the client times out, although it is instantiated. Subsequent calls to the client would then fail. Is there an API to get the status of the client?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 3 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: @tkakantousis, pagination is handled in `CollectionsBuilder.java` which does it on a java stream. 
I was thinking about extending `AbstractFacade` but didn't know which class would be the type parameter since `KafkaFacade` deals with multiple tables?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 2 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: there's no API for status

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 3 above)
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: We can omit the type parameter, it's not enforced.

Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: solved by splitting to multiple facades

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 2 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: it seems like it's taking care of timeouts. in [1] it sets retries and reconnects settings

[1] https://github.com/apache/kafka/blob/ac385c4c3a770728848438f28f4acb8854ffc868/clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java

Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: successfully created topics after the AdminClient times out (the default is 5 min), meaning it recreates after time out.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java [hunk 8]
```diff
@@ -96,197 +95,143 @@
...
-        "Could not find project: " + projectId);
-    }
-    kafkaFacade.shareTopic(project, topicName, projectId);
-
-    AclDTO aclDto = new AclDTO(projectShared.getName(),
-            Settings.KAFKA_ACL_WILDCARD,
-            "allow", Settings.KAFKA_ACL_WILDCARD, Settings.KAFKA_ACL_WILDCARD,
-            Settings.KAFKA_ACL_WILDCARD);
-    kafkaFacade.addAclsToTopic(topicName, project.getId(), aclDto);
     json.setSuccessMessage("The topic has been shared.");
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: @kai-chi in v2 we don use successMessage. That means all the code in the front-end that uses that needs to be refactored.

Comment (tkakantousis) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: In general, please stick with the codebase of Jobs/Executions (resource,controllers, test specs) as a guide for the Kafka API.

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java [hunk 9]
```diff
@@ -96,197 +95,143 @@
...
-    return noCacheResponse.getNoCacheResponseBuilder(Response.Status.OK).entity(json).build();
+    return Response.ok().entity(json).build();
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: @kai-chi Here we are creating a resource, which is a shared topic. The response code should be 201 and the response should contain the DTO of the created resource and the href to that resource. https://github.com/logicalclocks/hopsworks/blob/v0.10.0/hopsworks-api/src/main/java/io/hops/hopsworks/api/jobs/JobsResource.java#L182

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/topics/TopicsFilterBy.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
+/*
+ * This file is part of Hopsworks
+ * Copyright (C) 2018, Logical Clocks AB. All rights reserved
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/topics/TopicsFilterBy.java]: @kai-chi 2018 -> 2019.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java [hunk 11]
```diff
@@ -670,8 +183,8 @@ public void updateTopicAcl(Project project, String topicName, Integer aclId, Acl
     //remove previous acl
     em.remove(ta);
 
-    //add the new acls  
-    addAclsToTopic(topicName, project.getId(), aclDto);
+    //add the new acls
+    kafkaController.addAclsToTopic(topicName, project.getId(), aclDto);
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: @kai-chi KafkaController should not be referenced by the facade.

File: hopsworks-kmon/src/main/java/io/hops/hopsworks/kmon/user/AbstractFacade.java [hunk 12]
```diff
@@ -1,97 +0,0 @@
...
- * DAMAGES OR  OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
- * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
- */
-
-package io.hops.hopsworks.kmon.user;
-
-import java.util.List;
-import javax.persistence.EntityManager;
-
-public abstract class AbstractFacade<T> {
```
Comment (tkakantousis) [inline hopsworks-kmon/src/main/java/io/hops/hopsworks/kmon/user/AbstractFacade.java]: @kai-chi why is this removed?

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 11 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: correct, this will be refactored along with the ACL PR. All controller references will be removed PR by PR

File: hopsworks-kmon/src/main/java/io/hops/hopsworks/kmon/user/AbstractFacade.java (hunk 12 above)
Comment (kai-chi) [reply hopsworks-kmon/src/main/java/io/hops/hopsworks/kmon/user/AbstractFacade.java]: it's a phantom class. Should I bring it back?

Comment (tkakantousis) [reply hopsworks-kmon/src/main/java/io/hops/hopsworks/kmon/user/AbstractFacade.java]: yes, it should be removed in it's on kmon-related PR. We were thinking of removing kmon anyway.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 11 above)
Comment (tkakantousis) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: Yes but we can't merge code where the transactionless controller is used by the facade. If we need to merge topics and acls together, we'll do that then.

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java [hunk 13]
```diff
@@ -42,25 +42,51 @@
...
+import java.util.Comparator;
 import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
 import java.util.logging.Level;
 import java.util.logging.Logger;
+import java.util.stream.Collectors;
 import javax.ejb.EJB;
 import javax.ejb.Stateless;
```
Comment (tkakantousis) [inline hopsworks-common/src/main/java/io/hops/hopsworks/common/kafka/KafkaController.java]: @kai-chi transactions must be disabled for the controller

File: hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java (hunk 11 above)
Comment (kai-chi) [reply hopsworks-common/src/main/java/io/hops/hopsworks/common/dao/kafka/KafkaFacade.java]: I will remove the reference in this PR since it's the only one left.

File: hopsworks-IT/src/test/ruby/spec/kafka_spec.rb [hunk 14]
```diff
@@ -109,5 +117,227 @@
...
+      	    topic_name = create_topic(second_project.id)
+      		
+            # share topic with the first_project
+      	    share_topic(second_project, topic_name, project)
+      	    expect_status(201)
+      
+      	    # get all project topics for the first project
+      	    get_project_topics(project.id)
+            expect(json_body[:items].count).to eq(10)
+      	    # get all project topics for the first project
```
Comment (tkakantousis) [inline hopsworks-IT/src/test/ruby/spec/kafka_spec.rb]: @kai-chi fix comment

File: hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java [hunk 15]
```diff
@@ -363,13 +312,13 @@ public Response ValidateSchemaForTopics(SchemaDTO schemaData) throws KafkaExcept
     switch (kafkaFacade.schemaBackwardCompatibility(schemaData)) {
       case INVALID:
         json.setErrorMsg("schema is invalid");
-        return noCacheResponse.getNoCacheResponseBuilder(Response.Status.NOT_ACCEPTABLE).entity(json).build();
+        return Response.status(Response.Status.NOT_ACCEPTABLE).entity(json).build();
```
Comment (tkakantousis) [inline hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: @kai-chi we should throw a REST KafkaException and let the throwablemapper handle logging and sending the response.

Comment (kai-chi) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: @tkakantousis, this will be refactored in the schema PR. This change was simply done to get rid of noCacheResponse but changing it further is not  topics-related

Comment (tkakantousis) [reply hopsworks-api/src/main/java/io/hops/hopsworks/api/kafka/KafkaService.java]: Thanks
